### ISSUE TITLE
Documenting Signal vs Exceptions

### ISSUE THREAD

**Issue by drosehn** [crystal-lang/crystal#3987]
This is a minor documentation request, which came up in the middle of a long discussion about an unrelated issue.  I'm just making this suggestion into a separate github-issue so that it doesn't get lost:

A ruby programmer might be used to doing:
```cr
begin
    sleep sleep_sec
rescue  Interrupt    # this is how ruby would catch a control-C
    printf "\n#...Attn! at %s\n", Time.now.to_s("%T")
    exit 2
end
```
If you try that, crystal complains that `Interrupt` is not defined, which is fine.  The way to handle this in crystal is more like:
```cr
Signal::INT.trap do |x|
  puts "Handling INT"
end
sleep sleep_sec
```
When crystal gave me the compile-time error, my initial reaction was to look up the correct name for this exception, so I went to:
https://crystal-lang.org/api/Exception.html

I wouldn't expect any long tutorial there, but it might be nice to add a one-sentence pointer to say "Note that some events are handled by `Signal`, not by `Exceptions`".  However, I am pretty bad at writing documentation, and I'm never happy with whatever documentation I do write.  So I have no specific wording that I'd favor.  And if https://crystal-lang.org/api/Exception.html isn't the best place for that, then maybe over in:
https://crystal-lang.org/docs/syntax_and_semantics/exception_handling.html

**Comment by ysbaddaden** [comment]
I don't see any reason to document that. Signals must be handled for what they are. We don't have to document something because Ruby does raise exceptions sometimes —thought the correct way to handle this in Ruby is to use `Signal`, too.

**Comment by drosehn** [comment]
Speaking just to the ruby part:  Ruby does have `Signal`, e.g.:
```cr
Signal.trap("TERM") {
   do_cleanup_stuff
   exit 0
}
```
but it looks like I've only used it once in all of my scripts, and that was for something I copied from a gem.  But there's also the support in `rescue`, e.g.:
```cr
begin
    forever_do_aloop_of_stuff
rescue Interrupt
    #	Rescue is just to avoid the traceback caused when the user hits a
    #	ctrl-c when they want to exit the never-ending 'tail' operation.
    $stderr.printf "   ...Attn!\n"
rescue SignalException
    #   Also terminate the script cleanly when other signals are received,
    #	such as the SIGHUP (sometimes!) received at system shutdown, so
    #	that the final summary statistics will be printed.
    $stderr.printf "\n"
end
```
Note that doing it via `rescue` means that there is a well-defined scope that the interrupt-handler is in effect, without having the programmer handle pushing and popping what the current signal handler is.  I'm sure that all ruby does is set a flag in its own standard signal-handler as code enters or leaves a block.  At least that's how I would do it, instead of actually pushing and popping signal-handlers the way I would do in a C program.

A little googling shows that python programmers might be used to doing:
```cr
# SIGINT will normally raise a KeyboardInterrupt, just like any other Python call
try:
    socket.recv()
except KeyboardInterrupt:
    print("W: interrupt received, stopping…")
finally:
    # clean up
    socket.close()
    context.term()
```
Also note that I'm not asking anyone to implement `rescue Interrupt` or to even mention the word "ruby".  I'm just saying that new crystal programmers looking for exception-handling might not think to look at `Signal`.

**Comment by drosehn** [comment]
Actually, a bigger issue is to document signal itself.  The section about `Signal` in the API documentation is not very helpful.  It doesn't really say what signals are or in what contexts you would use them:
https://crystal-lang.org/api/Signal.html

And as near as I can see, the crystal-docs documentation has the word "signal" in only two places:
https://crystal-lang.org/docs/?q=signal

... and one of those is for the sentence "_The usual convention is to provide an alternative "question" method to signal that this variant of the method returns nil instead of raising._"  (a "question" method such as `$1?`, 😄 )  Which is to say, it has absolutely nothing to do with `Signal`.

**Comment by drosehn** [comment]
disclaimer: In my entire life I have written less than 100 lines of python.  I have no idea how common it is to use `except KeyboardInterrupt` in python!

**Comment by ysbaddaden** [comment]
I agree that `Signal` is lacking some API documentation, but we must document the language by itself (how do things work) not in contrast to how another language is doing it (differently); whatever the initial similarity with Ruby.

**Comment by spalladino** [comment]
Let's go with documenting the Signal API with some sample usage for now. I'm leaving the issue open should someone from the community wants to contribute with a PR to https://github.com/crystal-lang/crystal/blob/master/src/signal.cr with the documentation for https://crystal-lang.org/api/0.20.5/Signal.html.

**Comment by drosehn** [comment]
I'm used to working with `signal()`s in C, and a little bit in ruby, so I would want to do more work with them in crystal before trying to write decent documentation.  Here's one simple example program which tries to illustrate some aspects of `Signal` in crystal:
```cr
#
#   crystal build TestSignals.cr -o /tmp/TestSignals && /tmp/TestSignals
#

#   ... Make sure we avoid the EOF-on-STDIO bug in crystal, CF:
#       https://github.com/crystal-lang/crystal/issues/3674
#       https://github.com/crystal-lang/crystal/issues/2713
STDIN.blocking = true

macro do_sleep(num)
    sleep_sec = {{ num }}
    sleep_cnt += 1
    printf "Starting %d-second sleep at %s\n", sleep_sec, Time.now.to_s("%T")
    sleep sleep_sec
    printf "Finished 'sleep' #%d.\n", sleep_cnt
end

sleep_sec = 0
sleep_cnt = 0
int_cnt = 0
begin
    Signal::INT.trap do |x|
	#   This is the code which will be executed when a Ctrl-C
	#   is entered by the user.
	#   The parameter 'x' indicates which signal was captured,
	#   which is useful when trapping multiple signals in a
	#   single code-block.
	printf "\n#...Trapped INT! at %s, x=%s\n", Time.now.to_s("%T"),
	    x.inspect
	int_cnt += 1
	#   You could 'exit()' the program here, but this example
	#   will simply return to the interrupted 'sleep'.
    end

    do_sleep(10)
end

#   Note that the 'trap' is still in effect, outside the begin/end.
do_sleep(15)

#   Here is how to switch back to the default handler:
Signal::INT.reset
do_sleep(20)

printf "Caught %d ctrl-C's.\n", int_cnt
```
although I will say that the C-programmer in me finds it odd that the `sleep` picks up where it left off at the time of the interrupt.  I don't see any easy way to change that behavior, however I have not spent much time learning about `Signal`.  I'd also like to provide an example of trapping `HUP`, and maybe a few other often-used signals.  But I don't know much about using them in crystal yet!

**Comment by drosehn** [comment]
There are other things I'd like to illustrate, such has how to push/pop interrupt handlers.  If the second `Signal::INT.trap` wanted to switch back to "whatever was in effect before", without knowing what that was, how can it do that?

**Comment by drosehn** [comment]
Also, the global nature of `Signal.trap` suggests to me that maybe one example should have the trap set up in a separate module, so that code could change the way the `trap` routine behaved, without having to constantly push/pop different interrupt handlers.  For instance, some sections of code might want to `exit` when an interrupt comes in, and other sections might want the interrupt handler to log some warning message but then return to whatever code was interrupted.

I suspect I need to learn about crystal fibers before I'd understand the right way to do that.

**Comment by RX14** [comment]
@drosehn If you consider that there can only be one signal handler for a single process, and that crystal fibers can interleave arbitrarily, you'll find that being able to trap a Ctrl-C in a single block of code is "magic" provided by ruby, enabled by it's global interpreter lock. Since crystal will eventually be multi-core, able to execute multiple fibers in parallel, providing this magic will break down.

**Comment by spalladino** [comment]
Hey @drosehn, if you are interested in tackling the issue, I'd suggest going for a far simpler example, just showing a single usage of `Signal::INT.trap`, and explaining that the code is similar for handling other signals. I'd avoid multiple sleeps, counters, resets at least on a first example. 

As for pushing/popping handlers, I'm not sure there is a direct way to do it. Looking at [the code](https://github.com/crystal-lang/crystal/blob/1f3e8b0e742b55c1feb5584dc932e87034365f48/src/event/signal_handler.cr#L64), it would seem that registering a new handler simply overwrites the previous one, so implementing a stack would require some additional work. 

If you are interested in implementing something like that, it definitely sounds interesting, but I'd keep it in a separate shard, and maintain the std docs with a vanilla example.

Thanks!

**Comment by drosehn** [comment]
I do need to learn about fibers before attempting a good writeup.  The multiple sleeps are only there to help illustrate some aspects of how `Signal.trap` works, and I do think that information would be helpful to someone new to `Signal`s.

**Comment by ysbaddaden** [comment]
In the context of Crystal, Signal should only be considered for listening to what the OS wants to tell us, that is the most common scenarios: trap SIGINT andSIGTERM, how to ignore some signals, and maybe how to send one. Timers, alarms and other advanced and specialized topics shouldn't be documented by Crystal. If a developer really needs them, the signal(7) man page should be more than enough.

A note should detail that signals are received and handled by Crystal, then the registered handlers will be processed within the event loop, so:

- signals don't block the event loop (too much);
- signals may be handled in parallel;
- there are no limitation to what functionality is accessible within signal handlers (no whitelist of signal-safe functions).

**Comment by asterite** [comment]
`LibC.signal` does execute immediately, but then our signal handler (which was accepted via a PR and not carefully reviewed) uses pipes and other stuff... that's why the signal at the user level isn't executed immediately. I don't know if there's anything preventing this from working like that, but I think it should. For example if you hit CTRL+C while running specs it should immediately abort, but now it's not the case. I didn't know why it doesn't, it previously did, but it seems that after we accepted that PR it stopped working "well". I'd like to fix that if possible.

**Comment by ysbaddaden** [comment]
A signal handler can be interrupted by another signal, so functions called within the handler must be async-safe, and there is only a limited list of such functions; see the "Async-signal-safe functions" section of [signal(7)](http://man7.org/linux/man-pages/man7/signal.7.html). You can't allocate memory for example.

Some developers recommend to only trigger a boolean within the signal handler and nothing more; then your main loop or a thread will act upon the change; that may be a little extreme. We rely on the popular pipe trick for now: registered handlers do nothing more than push the signal to the pipe; the event loop will then resume a pending fiber, and it's now safe to do anything within the handler. Bonus: the event loop won't be stopped for too long.

I don't know how to fix the delay induced by handling the signal through the event loop, though. Maybe we should look at libevent which implements this solution. Maybe it would behave better?

**Comment by drosehn** [comment]
Yeah, I mean to ask whether crystal was doing some special magic such that the call to `printf` in my example would be okay, but I get the feeling that I'm just annoying people with my interest in systems-programming level topics.  There are several things about the implementation of `Signal` in crystal that seem odd to me, coming from a C background.  But I haven't taken the time to understand what crystal's code is doing.

The more I tried to do things with the current `Signal` in crystal, the more I wondered if all the headaches of using `signal()` (the C-level API) is the whole reason that ruby and python decided to implement stuff like `rescue Interrupt`.  That way all the messy low-level parts are handled by language's runtime support, making it much less likely for programmers to do stupid things (such as using `printf` in a signal handler!).

However, I just spent the last four hours shoveling snow out of my driveway, so I'm going to need some sleep before I try to say anything intelligent!

**Comment by ysbaddaden** [comment]
The problem of going through the event loop is that a busy loop has the consequence of never processing a signal. Stupid example follows: the handler is never run and the program never exits.

```crystal
Signal::INT.trap { puts "gotcha"; exit }
loop { |i| i += 1 }
```

  1. signals should be processed;
  2. in a timely manner;
  3. without blocking the event loop.

I think this promotes to a bug:

@drosehn low-level stuff is very interesting but doesn't belong to the Crystal documentation that shouldn't have to deal with theses tiny details. They would make great articles, though!

**Comment by drosehn** [comment]
@RX14 is giving a quick overview on fibers on the gitter channel, so that's helping me understand what's going on.

**Comment by drosehn** [comment]
Looking over what I wrote earlier, let me correct something that I phrased badly.  I wrote:
> I'd also like to provide an example of trapping `HUP`, and maybe a few other often-used signals. But I don't know much about using them in crystal yet!

First off, I meant "produce" instead of "provide", and secondly the sentence is still very misleading if I had used "produce".  Re-read that as:

I'd like to write an example of trapping `HUP`, and maybe a few other often-used signals.  I would do this for myself, to make sure I understand how `Signal` works in crystal.  And based on what I find out from those examples, I'll be confident that I will know what needs to be documented.

... but those examples are meant for my learning.  I didn't mean that I expected all those examples to be **in** the documentation.  Or at least, not unless every example happens to show something unusual about crystal's `Signal` as compared to the Unix-standard `signal()` routine.

**Comment by drosehn** [comment]
> @drosehn low-level stuff is very interesting but doesn't belong to the Crystal documentation that shouldn't have to deal with theses tiny details. They would make great articles, though!

This comment may have been a reaction to my badly-phrased sentence, but if it was not:

If the `Signal` module in crystal behaves significantly different than the more standard `signal()` API, then the documentation of that unique behavior has to be in crystal.  It is not helpful to tell users *"Please refer to the standard system documentation for all the tiny low-level tiny details about signals"*, when the system documentation is going to describe behavior which is very different from what they will get from `Signal`.  That's my opinion at least, even if it means that crystal's documentation needs to get into annoying minutia.

**Comment by drosehn** [comment]
But for now, I think my position is that we probably should wait before doing any detailed documentation of `Signal`, even though I'm the guy who asked for that.

Now that I have some idea of how fibers work, I suspect that we'll find out that crystal's support for `signal()` needs to be re-done in some significant way, especially once crystal has full support for parallel-processing (running on multiple CPU's), in addition to the current implementation of fibers which is limited to just cooperative-multitasking.

**Comment by shelvacu** [comment]
Signal trapping is badly undocumented. I only understood it once I looked at this thread. A simple practical example at the top of the [Signal page](https://crystal-lang.org/api/0.22.0/Signal.html) would be great, something like:

```crystal
Signal::USR1.trap do
  puts "Received SIGUSR1"
end
```

**Comment by straight-shoota** [comment]
This has already been added in #4587 and the documentation page will be updated with the next release.

**Comment by RX14** [comment]
Closed in #4587.
